### PR TITLE
Slide when colliding with Colliders

### PR DESCRIPTION
**Pull request by psykzz**
Currently, if you collide with a wall moving diagonally you will stop hard.
This will allow you to slide against the wall if there is the space.

### CODE REVIEW

File: SS14.Client.GameObjects/EntitySystems/TransformSystem.cs [hunk 1]
```diff
@@ -21,6 +21,37 @@ public TransformSystem(EntityManager em, EntitySystemManager esm)
             EntityQuery.Exclusionset.Add(typeof(SlaveMoverComponent));
         }
 
+        private Vector2f? calculateNewPosition(Entity entity, Vector2f newPosition, TransformComponent transform)
```
Comment (ComicIronic) [inline SS14.Client.GameObjects/EntitySystems/TransformSystem.cs:24]: Correct me if I'm wrong: the `?` makes this nullable?

Comment (psykzz) [reply SS14.Client.GameObjects/EntitySystems/TransformSystem.cs:24]: That's right

**Review by PJB3005 — APPROVED** [review]

**Review by d34d10cc — APPROVED** [review]

File: SS14.Client.GameObjects/EntitySystems/TransformSystem.cs [hunk 2]
```diff
@@ -101,16 +132,12 @@ public override void Update(float frametime)
...
-                            var collider = entity.GetComponent<ColliderComponent>(ComponentFamily.Collider);
-                            bool collided = collider.TryCollision(newPosition - transform.Position, true);
-                            if (!collided)
-                                doTranslate = true;
-                            else
+                            Vector2f? _newPosition = calculateNewPosition(entity, newPosition, transform);
+                            if (_newPosition != null)
                             {
-                                //Debugger.Break();
+                                newPosition = (Vector2f) _newPosition;
```
Comment (ComicIronic) [inline SS14.Client.GameObjects/EntitySystems/TransformSystem.cs]: I'm surprised casting works, but maybe `_newPosition.Value` is better - it removes the runtime check.

File: SS14.Client.GameObjects/EntitySystems/TransformSystem.cs (hunk 1 above)
Comment (RemieRichards) [reply SS14.Client.GameObjects/EntitySystems/TransformSystem.cs:24]: Why have you made it nullable? that's not required to return null, and you're not using anything specific to nullables, just checking ```!= null``` when you use the value.

also @ComicIronic yes, ```T?``` is sugar for ```Nullable<T>```

Comment (PJB3005) [reply SS14.Client.GameObjects/EntitySystems/TransformSystem.cs:24]: @RemieRichards it is for structs I believe.

Comment (PJB3005) [reply SS14.Client.GameObjects/EntitySystems/TransformSystem.cs:24]: afbeelding

yep

Comment (RemieRichards) [reply SS14.Client.GameObjects/EntitySystems/TransformSystem.cs:24]: interesting that classes are nullable but structs aren't. guess it's the whole "struct smells like an object but is actually a value" thing.

Comment (ComicIronic) [reply SS14.Client.GameObjects/EntitySystems/TransformSystem.cs:24]: It's also a syntax error to make an object type nullable, I believe.

File: SS14.Client.GameObjects/EntitySystems/TransformSystem.cs (hunk 2 above)
Comment (psykzz) [reply SS14.Client.GameObjects/EntitySystems/TransformSystem.cs]: Im good with that change, I don't have an environment here at work, so i'll make the change at home later.

File: SS14.Client.GameObjects/EntitySystems/TransformSystem.cs (hunk 1 above)
Comment (PJB3005) [reply SS14.Client.GameObjects/EntitySystems/TransformSystem.cs:24]: Structs are value types so they're presented in memory literally in the same way as things like integers IIRC, whereas classes are references types and managed by references. It makes sense.

Comment (RemieRichards) [reply SS14.Client.GameObjects/EntitySystems/TransformSystem.cs:24]: @PJB3005 
>guess it's the whole "struct smells like an object but is actually a value" thing.

but yeah.

**Comment by psykzz** [comment]
Ready for merge.

**Review by ComicIronic — APPROVED** [review]

**Pull request merged**